winch/codegen/src/codegen/call.rs

```diff
@@ -0,0 +1,144 @@
+use super::CodeGenContext;
+use crate::{
+    abi::{align_to, ABIArg, ABIResult, ABISig, ABI},
+    masm::{MacroAssembler, OperandSize},
+    reg::Reg,
+    stack::Val,
+};
+
+/// All the information needed to emit a function call.
+pub(crate) struct FnCall<'a> {
+    /// The total stack space needed for the callee arguments on the
+    /// stack, including any adjustments to the function's frame and
+    /// aligned to to the required ABI alignment.
+    pub total_arg_stack_space: u32,
+    /// The stack space consumed by spilling registers placed as
+    /// function arguments in the value stack.  This gets implicitly
+    /// allocated by saving any live registers in the value stack, but
+    /// we track it to make sure to reclaim that space once the
+    /// function call ends.
+    pub spill_stack_space: u32,
+    /// The ABI-specific signature of the callee.
+    pub abi_sig: &'a ABISig,
+    /// The stack pointer offset prior to preparing and emitting the
+    /// call. This is tracked to assert the position of the stack
+    /// pointer after the call has finished.
+    sp_offset_at_callsite: u32,
+}
+
+impl<'a> FnCall<'a> {
+    /// Allocate and setup a new function call.
+    ///
+    /// The setup process, will first save all the live registers in
+    /// the value stack, tracking down those spilled for the function
+    /// arguments.  Then, it will calculate any adjustments needed to
+    /// ensure the alignment of the caller's frame.  It's important to
+    /// note that the order of operations in the setup is important,
+    /// as we want to calculate any adjustments to the caller's frame,
+    /// after having saved any live registers, so that we can account
+    /// for any pushes generated by register spilling.
+    pub fn new<A: ABI, M: MacroAssembler>(
+        abi: &A,
+        callee_sig: &'a ABISig,
+        context: &mut CodeGenContext,
+        masm: &mut M,
+    ) -> Self {
+        let stack = &context.stack;
+        let arg_stack_space = callee_sig.stack_bytes;
+        let callee_params = &callee_sig.params;
+        let sp_offset_at_callsite = masm.sp_offset();
+
+        let spilled_regs = match callee_params.len() {
+            0 => {
+                let _ = context.spill_regs_in(masm, ..);
```

Comment:
Does this need to be a special case? It seems that if instead the `spill_regs_in` method accepted an empty range, this case would work with the code below...

Reply:
I was not able to find a better way to do this. In this case we still want to spill everything, to save any live registers, but we are no interested in the count returned by `spill_regs_and_count_memory_in`, since they are not part of the function arguments so we can't claim that memory after the function call, so that's why I'm handling `0` in this way.